src/python/pants/option/option_types.py

```diff
@@ -288,6 +288,45 @@ def __new__(
         )
 
 
+class DictOption(_OptionBase["dict[str, _ValueT]"], Generic[_ValueT]):
+    """A dictionary option mapping strings to client-provided `_ValueT`.
+
+    If you provide a `default` parameter, the `_ValueT` type parameter will be inferred from the
+    type of the values in the default. Otherwise, you'll need to provide `_ValueT` if you want a
+    non-`Any` type.
+
+    E.g.
+        # Explicit
+        DictOption[str](...)  # property type is `dict[str, str]`
+        DictOption[Any](..., default=dict(key="val"))  # property type is `dict[str, Any]`
+        # Implicit
+        DictOption(...)  # property type is `dict[str, Any]`
+        DictOption(..., default={"key": "val"})  # property type is `dict[str, str]`
+        DictOption(..., default=dict(key="val"))  # property type is `dict[str, str]`
+        DictOption(..., default=dict(key=1))  # property type is `dict[str, int]`
+        DictOption(..., default=dict(key1=1, key2="str"))  # property type is `dict[str, Any]`
```

Comment:
Maybe use `{"": ""}` syntax for both these? It's more familiar to most Python devs imo.

Reply:
I actually prefer `dict()` when the keys are valid identifiers :sweat_smile: Looks nicer with syntax highlighting and usually plays in IDE's autocompletion

Reply:
John agrees too. I'm totally fine with using `dict()` in actual call sites, only thinking for the docstring. But nbd either way